altair/utils/_vegafusion_data.py

```diff
@@ -0,0 +1,199 @@
+import sys
+from toolz import curried
+import uuid
+from weakref import WeakValueDictionary
+
+from typing import Union, Dict, Set, MutableMapping
+
+if sys.version_info >= (3, 8):
+    from typing import TypedDict, Final
+else:
+    from typing_extensions import TypedDict, Final
+
+from altair.utils.core import _DataFrameLike
+from altair.utils.data import _DataType, _ToValuesReturnType, MaxRowsError
+from altair.vegalite.data import default_data_transformer
+
+# Temporary storage for dataframes that have been extracted
+# from charts by the vegafusion data transformer. Use a WeakValueDictionary
+# rather than a dict so that the Python interpreter is free to garbage
+# collect the stored DataFrames.
+extracted_inline_tables: MutableMapping[str, _DataFrameLike] = WeakValueDictionary()
+
+# Special URL prefix that VegaFusion uses to denote that a
+# dataset in a Vega spec corresponds to an entry in the `inline_datasets`
+# kwarg of vf.runtime.pre_transform_spec().
+VEGAFUSION_PREFIX: Final = "vegafusion+dataset://"
+
+
+class _ToVegaFusionReturnUrlDict(TypedDict):
+    url: str
+
+
+@curried.curry
+def vegafusion_data_transformer(
+    data: _DataType, max_rows: int = 100000
+) -> Union[_ToVegaFusionReturnUrlDict, _ToValuesReturnType]:
+    """VegaFusion Data Transformer"""
+    if hasattr(data, "__geo_interface__"):
```

Comment:
I think this escape hatch for geo data is not yet implemented in the other parts of this PR such as the rendering and saving of charts. The following example from the Altair docs raises a `PanicError`:

```python
import altair as alt
import geopandas as gpd

alt.data_transformers.enable("vegafusion")

url = "https://naciscdn.org/naturalearth/110m/cultural/ne_110m_admin_0_countries.zip"
gdf_ne = gpd.read_file(url)  # zipped shapefile
gdf_ne = gdf_ne[["NAME", "CONTINENT", "POP_EST", 'geometry']]

alt.Chart(gdf_ne).mark_geoshape()
```

Somehow `using_vegafusion` needs to be aware of this distinction as well?

Reply:
Thanks for catching and flagging this. The issue here is that the JSON-to-Arrow logic VegaFusion uses is choking on the inline geojson data. I think VegaFusion needs to be better about gracefully passing through complex JSON that Arrow fails to parse.

Reply:
Tracking in https://github.com/hex-inc/vegafusion/issues/352 and will be fixed in VegaFusion 1.4.0-rc2

Reply:
Great to hear, thanks!